core/src/main/java/org/infinispan/interceptors/locking/ClusteringDependentLogic.java

```diff
@@ -508,47 +584,79 @@ protected void commitSingleEntry(CacheEntry entry, Metadata metadata, FlagAffect
+                  removed = entry.isRemoved();
+                  if (removed && entry instanceof MVCCEntry) {
+                     expired = ((MVCCEntry) entry).isExpired();
+                  }
                }
-            }
 
-            if (doCommit) {
+               // TODO use value from the entry
                InternalCacheEntry previousEntry = dataContainer.peek(entry.getKey());
                Object previousValue = null;
                Metadata previousMetadata = null;
                if (previousEntry != null) {
                   previousValue = previousEntry.getValue();
                   previousMetadata = previousEntry.getMetadata();
                }
-               commitManager.commit(entry, metadata, trackFlag, l1Invalidation);
-               if (!isForeignOwned) {
-                  notifyCommitEntry(created, removed, expired, entry, ctx, command, previousValue, previousMetadata);
+               if (isL1Write && previousEntry != null && !previousEntry.isL1Entry()) {
+                  // don't overwrite non-L1 entry with L1 (e.g. when originator == backup
+                  // and therefore we have two contexts on one node)
+               } else {
+                  commitManager.commit(entry, metadata, trackFlag, l1Invalidation);
+                  if (doCommit.isLocal()) {
+                     notifyCommitEntry(created, removed, expired, entry, ctx, command, previousValue, previousMetadata);
+                  }
                }
-            } else
-               entry.rollback();
-
+            }
          } finally {
             stateTransferLock.releaseSharedTopologyLock();
          }
       }
 
       @Override
+      public Commit shouldCommit(FlagAffectedCommand command, InvocationContext ctx, Object key, boolean removed) {
+         // ignore locality for removals, even if skipOwnershipCheck is not true
+         if (command != null && command.hasFlag(Flag.SKIP_OWNERSHIP_CHECK)) {
+            return Commit.COMMIT_LOCAL;
+         }
+         boolean transactional = ctx.isInTxScope() && (command == null || !command.hasFlag(Flag.PUT_FOR_EXTERNAL_READ));
+         // When a command is local-mode, it does not get written by replicating origin -> primary -> backup but
+         // when origin == backup it's written right from the original context
+         // ClearCommand is also broadcast to all nodes from originator, and on originator it should remove entries
+         // for which this node is backup owner even though it did not get the removal from primary owner.
+         if (transactional || !ctx.isOriginLocal() || (command != null &&
+               (command.hasFlag(Flag.CACHE_MODE_LOCAL) || command instanceof ClearCommand))) {
+            // During ST, entries whose ownership is lost are invalidated by InvalidateCommand
+            // without any flags, and at that point we're no longer owners - the only information is that
+            // the origin is not local and the entry is removed.
```

Comment:
`InvalidateCommand` is never replicated, so it doesn't really need `CACHE_MODE_LOCAL`, but we can add it if it simplifies things.